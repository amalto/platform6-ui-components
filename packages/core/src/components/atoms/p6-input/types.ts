export enum P6InputType {
  email = 0,
  number = 1,
  password = 2,
  search = 3,
  tel = 4,
  text = 5,
  url = 6,
}
