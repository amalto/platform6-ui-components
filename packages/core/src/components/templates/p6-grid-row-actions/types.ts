import { IconName, IconPrefix } from '@fortawesome/fontawesome-svg-core';
import { EventEmitter } from '@stencil/core';
import { Mode } from '../../../shared/types';
import { RowId } from '../../organisms/p6-grid/core/entities';

export interface ActionsOptions {
  create: { disabled: boolean; canUse: boolean };
  view: { disabled: boolean; canUse: boolean };
  edit: { disabled: boolean; canUse: boolean };
  rename: { disabled: boolean; canUse: boolean };
  delete: { disabled: boolean; canUse: boolean };
  export: { disabled: boolean; canUse: boolean };
  [action: string]: { disabled: boolean; canUse: boolean };
}

export interface Action {
  id: keyof ActionsOptions;
  title: string;
  icon: IconName;
  iconPrefix?: IconPrefix;
  mode?: Mode;
  emitter: () => EventEmitter<{ rowId: RowId }> | undefined;
}
