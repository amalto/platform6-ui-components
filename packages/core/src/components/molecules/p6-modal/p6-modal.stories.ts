import { library } from '@fortawesome/fontawesome-svg-core';
import { faHeart, faReply, faRetweet } from '@fortawesome/free-solid-svg-icons';
import { html, TemplateResult } from 'lit-html';

export default {
  title: 'Molecules/Modal',
  component: 'p6-modal',
  argTypes: {
    hasCard: { control: 'boolean' },
    open: false,
  },
};

library.add(faReply, faRetweet, faHeart);

export const Default = ({ hasCard, open }: { hasCard: boolean; open: boolean }): TemplateResult => {
  return html`
    <p6-modal open="${open}" hasCard="'${hasCard}'">
      <article class="media">
        <div class="media-left">
          <figure class="image is-64x64">
            <img src="https://bulma.io/images/placeholders/128x128.png" alt="Image" />
          </figure>
        </div>
        <div class="media-content">
          <div class="content">
            <p>
              <strong>John Smith</strong> <small>@johnsmith</small> <small>31m</small>
              <br />
              Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean efficitur sit amet massa fringilla egestas. Nullam condimentum luctus turpis.
            </p>
          </div>
          <nav class="level is-mobile">
            <div class="level-left">
              <a class="level-item" aria-label="reply">
                <p6-icon name="reply" size="0" aria-hidden="true" />
              </a>
              <a class="level-item" aria-label="retweet">
                <p6-icon name="retweet" size="0" aria-hidden="true" />
              </a>
              <a class="level-item" aria-label="like">
                <p6-icon name="heart" size="0" aria-hidden="true" />
              </a>
            </div>
          </nav>
        </div>
      </article>
    </p6-modal>
  `;
};

Default.args = {
  hasCard: false,
  open: true,
};

export const ModalCard = ({ hasCard, open }: { hasCard: boolean; open: boolean }): TemplateResult => {
  return html`
    <p6-modal open="${open}" has-card="${hasCard}">
      <p slot="head">Modal title</p>

      <div class="section-content">
        <div class="content">
          <h1>Hello World</h1>
          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla accumsan, metus ultrices eleifend gravida, nulla nunc varius lectus, nec rutrum justo nibh eu lectus. Ut
            vulputate semper dui. Fusce erat odio, sollicitudin vel erat vel, interdum mattis neque.
          </p>
          <h2>Second level</h2>
          <p>
            Curabitur accumsan turpis pharetra <strong>augue tincidunt</strong> blandit. Quisque condimentum maximus mi, sit amet commodo arcu rutrum id. Proin pretium urna vel
            cursus venenatis. Suspendisse potenti. Etiam mattis sem rhoncus lacus dapibus facilisis. Donec at dignissim dui. Ut et neque nisl.
          </p>
          <ul>
            <li>In fermentum leo eu lectus mollis, quis dictum mi aliquet.</li>
            <li>Morbi eu nulla lobortis, lobortis est in, fringilla felis.</li>
            <li>Aliquam nec felis in sapien venenatis viverra fermentum nec lectus.</li>
            <li>Ut non enim metus.</li>
          </ul>
          <h3>Third level</h3>
          <p>
            Quisque ante lacus, malesuada ac auctor vitae, congue <a href="#">non ante</a>. Phasellus lacus ex, semper ac tortor nec, fringilla condimentum orci. Fusce eu rutrum
            tellus.
          </p>
          <ol>
            <li>Donec blandit a lorem id convallis.</li>
            <li>Cras gravida arcu at diam gravida gravida.</li>
            <li>Integer in volutpat libero.</li>
            <li>Donec a diam tellus.</li>
            <li>Aenean nec tortor orci.</li>
            <li>Quisque aliquam cursus urna, non bibendum massa viverra eget.</li>
            <li>Vivamus maximus ultricies pulvinar.</li>
          </ol>
          <blockquote>Ut venenatis, nisl scelerisque sollicitudin fermentum, quam libero hendrerit ipsum, ut blandit est tellus sit amet turpis.</blockquote>
          <p>
            Quisque at semper enim, eu hendrerit odio. Etiam auctor nisl et
            <em>justo sodales</em> elementum. Maecenas ultrices lacus quis neque consectetur, et lobortis nisi molestie.
          </p>
          <p>Sed sagittis enim ac tortor maximus rutrum. Nulla facilisi. Donec mattis vulputate risus in luctus. Maecenas vestibulum interdum commodo.</p>
          <p>Suspendisse egestas sapien non felis placerat elementum. Morbi tortor nisl, suscipit sed mi sit amet, mollis malesuada nulla. Nulla facilisi. Nullam ac erat ante.</p>
          <h4>Fourth level</h4>
          <p>Nulla efficitur eleifend nisi, sit amet bibendum sapien fringilla ac. Mauris euismod metus a tellus laoreet, at elementum ex efficitur.</p>
          <p>
            Maecenas eleifend sollicitudin dui, faucibus sollicitudin augue cursus non. Ut finibus eleifend arcu ut vehicula. Mauris eu est maximus est porta condimentum in eu
            justo. Nulla id iaculis sapien.
          </p>
          <p>
            Phasellus porttitor enim id metus volutpat ultricies. Ut nisi nunc, blandit sed dapibus at, vestibulum in felis. Etiam iaculis lorem ac nibh bibendum rhoncus. Nam
            interdum efficitur ligula sit amet ullamcorper. Etiam tristique, leo vitae porta faucibus, mi lacus laoreet metus, at cursus leo est vel tellus. Sed ac posuere est.
            Nunc ultricies nunc neque, vitae ultricies ex sodales quis. Aliquam eu nibh in libero accumsan pulvinar. Nullam nec nisl placerat, pretium metus vel, euismod ipsum.
            Proin tempor cursus nisl vel condimentum. Nam pharetra varius metus non pellentesque.
          </p>
          <h5>Fifth level</h5>
          <p>
            Aliquam sagittis rhoncus vulputate. Cras non luctus sem, sed tincidunt ligula. Vestibulum at nunc elit. Praesent aliquet ligula mi, in luctus elit volutpat porta.
            Phasellus molestie diam vel nisi sodales, a eleifend augue laoreet. Sed nec eleifend justo. Nam et sollicitudin odio.
          </p>
          <h6>Sixth level</h6>
          <p>
            Cras in nibh lacinia, venenatis nisi et, auctor urna. Donec pulvinar lacus sed diam dignissim, ut eleifend eros accumsan. Phasellus non tortor eros. Ut sed rutrum
            lacus. Etiam purus nunc, scelerisque quis enim vitae, malesuada ultrices turpis. Nunc vitae maximus purus, nec consectetur dui. Suspendisse euismod, elit vel rutrum
            commodo, ipsum tortor maximus dui, sed varius sapien odio vitae est. Etiam at cursus metus.
          </p>
        </div>
      </div>
      <footer slot="foot">
        <section>
          <p6-button mode="3">Save changes</p6-button>
          <p6-button>Cancel</p6-button>
        </section>
      </footer>
    </p6-modal>
  `;
};

ModalCard.args = {
  hasCard: true,
  open: true,
};
