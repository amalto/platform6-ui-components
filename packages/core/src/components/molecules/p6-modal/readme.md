# p6-modal

<!-- Auto Generated Below -->

## Properties

| Property  | Attribute  | Description                                    | Type      | Default |
| --------- | ---------- | ---------------------------------------------- | --------- | ------- |
| `hasCard` | `has-card` | Display a modal with a head, a body and a foot | `boolean` | `false` |
| `open`    | `open`     | Open the modal                                 | `boolean` | `false` |

---

_Built with [StencilJS](https://stenciljs.com/)_
