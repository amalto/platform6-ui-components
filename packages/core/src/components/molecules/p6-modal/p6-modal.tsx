import { Component, Fragment, h, Prop } from '@stencil/core';

@Component({
  tag: 'p6-modal',
  styleUrl: 'p6-modal.scss',
  shadow: true,
})
export class P6Modal {
  /**
   * Display a modal with a head, a body and a foot
   */
  @Prop() hasCard = false;

  /**
   * Open the modal
   */
  @Prop({ mutable: true, reflect: true }) open = false;

  render(): JSX.Element {
    const classList = {
      'modal': true,
      'is-active': this.open,
    };

    return (
      <div class={classList}>
        <div class="modal-background" />

        {this.hasCard ? this.renderCard() : this.renderContent()}
      </div>
    );
  }

  private closeHandler(): (event: Event) => void {
    return () => {
      this.open = false;
    };
  }

  private renderContent(): JSX.Element {
    return (
      <Fragment>
        <div class="modal-content">
          <slot />
        </div>

        <button slot="close" class="modal-close is-large" type="button" aria-label="close" onClick={this.closeHandler()} />
      </Fragment>
    );
  }

  private renderCard(): JSX.Element {
    return (
      <div class="modal-card">
        <header class="modal-card-head">
          <span class="modal-card-title">
            <slot name="head" />
          </span>
          <button class="modal-card-close" type="button" aria-label="close" onClick={this.closeHandler()} />
        </header>

        <section class="modal-card-body">
          <slot />
        </section>

        <slot name="foot" />
      </div>
    );
  }
}
