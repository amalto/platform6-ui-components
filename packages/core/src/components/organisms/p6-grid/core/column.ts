import { Alignment, Direction, SortOrder } from '../../../../shared/types';
import { throwError } from '../../../../utils/error';
import { Column, ColumnDefinition, ColumnId, DataItem } from './entities';

export const MIN_WIDTH = 75;
export const DEFAULT_WIDTH = MIN_WIDTH;
export const INC_WIDTH = 10;

export function defaultGetValue<T extends DataItem>(data: T, context: ColumnDefinition<T>): string {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'getValue' function for ${JSON.stringify(context)}`);
  }

  return `${data[context.field as keyof T]}`;
}

export function defaultSetValue<T extends DataItem>(value: string, data: T, context: ColumnDefinition<T>): T {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'setValue' function for ${JSON.stringify(context)}`);
  }

  return {
    ...data,
    [context.field as keyof T]: value,
  };
}

let columnIds = 0;
export function fromDefinition<T extends DataItem>(column: ColumnDefinition<T>): Column<T> {
  // eslint-disable-next-line no-plusplus
  const id = column.id || column.field || columnIds++;

  return {
    ...column,
    id: `col-${id}`,
    editable: column.editable || false,
    hidden: column.hidden || false,
    width: column.width || 100,
    align: column.align || Alignment.center,
    sortOrder: column.sortOrder || SortOrder.none,
    disableHeaderMenu: column.disableHeaderMenu || false,
    getValue: column.getValue || defaultGetValue,
    setValue: column.setValue || defaultSetValue,
  };
}

export function isHidden<T extends DataItem>(column: Column<T>): boolean {
  return !!column.hidden;
}

export function replaceColumn<T extends DataItem>(updatedColumn: Column<T>, allColumns: Column<T>[]): Column<T>[] {
  return allColumns.map(col => (col.id === updatedColumn.id ? updatedColumn : col));
}

export function move<T extends DataItem>(id: ColumnId, direction: Direction, columns: Column<T>[]): Column<T>[] {
  const from = columns.findIndex(col => col.id === id);
  const to = from + (direction === Direction.left ? -1 : 1);

  if (from < 0 || to < 0 || to >= columns.length) {
    return columns;
  }

  const result = columns.slice();
  const item = result.splice(from, 1);

  return result.slice(0, to).concat(item).concat(result.slice(to, columns.length));
}
