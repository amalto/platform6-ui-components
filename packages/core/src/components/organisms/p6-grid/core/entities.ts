import { EventEmitter, JSX } from '@stencil/core';
import { Alignment, Direction, SortOrder } from '../../../../shared/types';

export type ColumnId = string;

export type DataItem = Record<string, unknown>;

export interface ColumnDefinition<T extends DataItem> {
  id?: ColumnId;
  label: string;
  field?: string | keyof T;
  editable?: boolean;
  filtreable?: boolean;
  sortable?: boolean;
  hidden?: boolean;
  width?: number;
  align?: Alignment;
  color?: string;
  sortOrder?: SortOrder;
  disableHeaderMenu?: boolean;
  getValue?: (data: T, context: ColumnDefinition<T>) => string;
  setValue?: (value: string, data: T, context: ColumnDefinition<T>) => T;
  cellEditor?: (data: T, context: ColumnDefinition<T>) => JSX.Element & CellEditor<T>;
  cellRenderer?: (data: T, rowId: RowId, context: ColumnDefinition<T>) => string | HTMLElement;
}

export interface Column<T extends DataItem> extends ColumnDefinition<T> {
  id: ColumnId;
  editable: boolean;
  hidden: boolean;
  width: number;
  align: Alignment;
  sortOrder: SortOrder;
  disableHeaderMenu: boolean;
  getValue: (data: T, context: ColumnDefinition<T>) => string;
  setValue: (value: string, data: T, context: ColumnDefinition<T>) => T;
}

export interface CellEditor<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  value: string;

  p6GridCellValueChanged: EventEmitter<CellValueChangedDetail<T>>;
  p6GridEditingCell: EventEmitter<EditingCellDetail<T>>;
}
export type RowId = string;

export interface Row<T extends DataItem> {
  id: RowId;
  data: T;
}

interface ColumnEventDetail<T extends DataItem> {
  column: Column<T>;
}

export interface MoveColumnDetail<T extends DataItem> extends ColumnEventDetail<T> {
  direction: Direction;
}
export type ResizeColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type AlignColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type SortColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type ShowColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export interface CellValueChangedDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  oldValue: string;
}

export enum EditingCellStatus {
  Start,
  Stop,
  Cancel,
}

export interface EditingCellDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  status: EditingCellStatus;
}

export interface ResetDefinitionsDetail {
  reset: boolean;
}

export interface ShowOptionsDetail {
  visible: boolean;
}

export interface FilterRowsDetail {
  value: string;
}
