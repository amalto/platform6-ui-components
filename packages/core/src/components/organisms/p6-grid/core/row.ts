import { SortOrder } from '../../../../shared/types';
import { Column, DataItem, Row, RowId } from './entities';

let rowIds = 0;
export function fromData<T extends DataItem>(data: T): Row<T> {
  rowIds += 1;

  return {
    id: `row-${rowIds}`,
    data,
  };
}

export function compareRow<T extends DataItem>(column: Column<T> | undefined): (a: Row<T>, b: Row<T>) => number {
  const sortOrder = column?.sortOrder || SortOrder.none;
  const factor = sortOrder === SortOrder.desc ? -1 : 1;

  return (a: Row<T>, b: Row<T>): number => {
    if (column === undefined || column.sortable === false || sortOrder === SortOrder.none) {
      return 0;
    }

    const valueA = column.getValue(a.data, column);
    const valueB = column.getValue(b.data, column);

    return valueA.localeCompare(valueB) * factor;
  };
}

export function filterBy<T extends DataItem>(searchValue: string, columns: Column<T>[]): (row: Row<T>) => boolean {
  return row => columns.filter(col => col.filtreable !== false).some(col => col.getValue(row.data, col).includes(searchValue));
}

export function rangeSelectRow<T extends DataItem>(currentRowIdSelect: RowId, selectedRows: Set<RowId>, displayedRow: Row<T>[]): Set<RowId> {
  const currentSelectedIndex = displayedRow.findIndex(row => row.id === currentRowIdSelect);

  if (currentSelectedIndex === -1) {
    return selectedRows;
  }

  const previousRowIdSelected = Array.from(selectedRows)
    .reverse()
    .find(rowId => displayedRow.findIndex(row => row.id === rowId) > -1);

  if (previousRowIdSelected === undefined) {
    return new Set(selectedRows.add(currentRowIdSelect));
  }

  const previousSelectedIndex = displayedRow.findIndex(row => row.id === previousRowIdSelected);

  let [from, to] = [previousSelectedIndex, currentSelectedIndex];
  let excludedFromSlice = currentRowIdSelect;
  if (currentSelectedIndex < previousSelectedIndex) {
    [from, to] = [to, from];
    excludedFromSlice = previousRowIdSelected;
  }

  return displayedRow
    .slice(from, to)
    .reduce((selected, cur) => {
      return cur.id === undefined ? selected : selected.add(cur.id);
    }, new Set(selectedRows))
    .add(excludedFromSlice);
}

export function replaceRow<T extends DataItem>(updatedRow: Row<T>, allRows: Row<T>[]): Row<T>[] {
  return allRows.map(row => (row.id === updatedRow.id ? updatedRow : row));
}
