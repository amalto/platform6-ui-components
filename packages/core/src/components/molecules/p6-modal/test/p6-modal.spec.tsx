import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { Mode } from '../../../../shared/types';
import { P6Modal } from '../p6-modal';

describe('p6-modal', () => {
  it('renders a content modal', async () => {
    const page = await newSpecPage({
      components: [P6Modal],
      template: () => {
        <p6-modal open hasCard={false}>
          <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean efficitur sit amet massa fringilla egestas. Nullam condimentum luctus turpis.</p>
        </p6-modal>;
      },
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders a card modal', async () => {
    const page = await newSpecPage({
      components: [P6Modal],
      template: () => {
        <p6-modal open hasCard>
          <span slot="head">Modal title</span>

          <div class="content">
            <h1>Hello World</h1>
            <p>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla accumsan, metus ultrices eleifend gravida, nulla nunc varius lectus, nec rutrum justo nibh eu lectus.
              Ut vulputate semper dui. Fusce erat odio, sollicitudin vel erat vel, interdum mattis neque.
            </p>
          </div>

          <footer slot="foot">
            <section>
              <p6-button mode={Mode.primary}>Save changes</p6-button>
              <p6-button>Cancel</p6-button>
            </section>
          </footer>
        </p6-modal>;
      },
    });
    expect(page.root).toMatchSnapshot();
  });
});
