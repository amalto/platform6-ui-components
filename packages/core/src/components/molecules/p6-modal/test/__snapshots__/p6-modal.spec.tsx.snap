// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-modal renders a card modal 1`] = `null`;

exports[`p6-modal renders a content modal 1`] = `null`;
